### PR TITLE
Add multiview support to the clustered forward renderer

### PR DESCRIPTION
**Pull request by BastiaanOlij**
This is split off from #48207

As there are a number of things left to finish off and while we focus on making the mobile renderer work properly for VR first, implementing this logic in the clustered forward renderer was moved into this separate PR.

Things still left to do:
- Fix clustered culling logic to work with the stereo implementation (though it works atm but it can be made better)
- Fix volumetric fog and GI
- Fix various post effects
These can all be done in future PRs.

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/scene_forward_clustered.glsl [hunk 1]
```diff
@@ -391,10 +409,26 @@ layout(location = 8) in float dp_clip;
...
+#else // USE_MULTIVIEW
+// Set to zero, not supported in non stereo
+#define ViewIndex 0
+#endif //USE_MULTIVIEW
+
 //defines to keep compatibility with vertex
 
 #define world_matrix instances.data[instance_index].transform
+#ifdef USE_MULTIVIEW
 #define projection_matrix scene_data.projection_matrix
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/shaders/scene_forward_clustered.glsl:428]: Oops, this is the wrong way around... gottah fix.

**Comment by BastiaanOlij** [comment]
Rebased this now #48207 has been merged.

**Comment by BastiaanOlij** [comment]
I've changed this to ready for review, the core of stereoscopic rendering works for the clustered renderer, its only around the effects that things fail but these can be handled in separate PRs (we should raise issues for these), some of which are shared with the mobile renderer and still need to be enhanced for that renderer as well.

It doesn't make sense to have this linger around while some people may want to start experimenting with VR on the high end renderer.

**Comment by clayjohn** [comment]
Which effects are currently broken?

**Comment by BastiaanOlij** [comment]
> 
> 
> Which effects are currently broken?

Most of the post process effects like glow, dof, etc, because they haven't been adjusted for multiview. The right eye will likely use the left eye for input or something like that.

Also I believe sky in half-res or quarter-res mode will currently break in stereo because the asymmetric view frustums mean you can't use a single baked image.

**Comment by BastiaanOlij** [comment]
Just rebased but currently not working, something wrong with the code the shaders expect

**Comment by BastiaanOlij** [comment]
Ok, works again

**Comment by BastiaanOlij** [comment]
@JFonS Reduz asks specifically if you could check if the changes I make here aren't getting in the way of your changes to the shader variants as I'm introducing some more variations. It comes back to the discussion we had a few days ago where we're just getting so many combo's it's kind of stressing how we currently handle those.

**Comment by Calinou** [comment]
> Occlusion culling right now works on the combined view frustum, it's good enough but there are situations where things are occluded from one eye and not the other and they may not be shown as a result.

Can this be worked around by adding an extra margin to every culled object that's equivalent to the distance between the two eyes?

**Comment by BastiaanOlij** [comment]
> > Occlusion culling right now works on the combined view frustum, it's good enough but there are situations where things are occluded from one eye and not the other and they may not be shown as a result.
> 
> Can this be worked around by adding an extra AABB margin to every culled object that's equivalent to the distance between the two eyes? I'm not sure if we should rasterize the occlusion culling buffer twice, as this can use up a lot of CPU resources in complex scenes (especially at high framerates).

That's actually not such a bad idea... A few minor false negatives will easily out way preventing a few false positives. The eye does pick up very quickly when something starts flickering in one eye has it's half occluded.

**Review by JFonS — APPROVED** [review]
The shader/pipeline changes look good. If I can fix the final kinks on my rework we should be able to cut down on all those new `enum` values, but we can merge this as-is and I will rebase on top.

**Comment by BastiaanOlij** [comment]
@reduz would be good if you could give your thumbsup on this as well, I'd like to build my cleanup work on the render engine ontop of this instead of requiring to keep this PR up to date :)
Also easier for @JFonS I'm guessing seeing the changes he is working on to just take these changes along.

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by BastiaanOlij** [comment]
Hurray!

**Pull request merged**